### PR TITLE
ELM327 Logging Support

### PR DESCRIPTION
**Pull request by GTO2013**
Hi,
this is the first implementation of ELM327 OBD logging support. Only live data is supported at the moment, so you cannot read DTCs.
You can use this simulator for testing: https://github.com/Ircama/ELM327-emulator
It is disabled by default and replaces the default serial communication when it is enabled.

There are probably many things to improve on this, but I wanted to commit the current state before there are any merge conflicts. One of the things is that inside the logger definition, the transport id defines which protocol is used. Choices are:

```
Automatic
saeJ1850-PWM
saeJ1850-VPW
iso9141-2 
iso4230-4KWP-5 
iso14230-4KWP-Fast
iso15765-4CAN11-500
iso15765-4CAN29-500
iso15765-4CAN11-250
iso15765-4CAN29-250
isoJ1939-250
```

The default iso15765 will be converted to iso15765-4CAN11-500. Not 100% happy with this, but I didnt see a good alternative, since the baudrate is inside the protocol element and not the transport element. So you cannot set different baudrates for different transport protcols and have to define it in the id.

Also: When searching for (non existing) J2534 drivers on windows RomRaider would just throw a registry error, it will print something more useful now. I also cleaned up an error inside the Dyno Feature, which also threw an error everytime you started RomRaider. I fixed a similar issue with the MTS plugin. I created a PluginNotInstalledException there which can be thrown if the plugin is not installed, which is probably the case for most users.

If there are any changes needed, please keep this request open so I dont have to recreate a pull request every time.

elmLogging

Cheers! :)

### CODE REVIEW

File: .classpath [hunk 1]
```diff
@@ -20,9 +20,10 @@
 	<classpathentry kind="lib" path="lib/common/cmutimelex.jar"/>
 	<classpathentry kind="lib" path="lib/common/en_us.jar"/>
 	<classpathentry kind="lib" path="lib/common/freetts.jar"/>
-	<classpathentry kind="lib" path="lib/common/RXTXcomm.jar">
+	<classpathentry exported="true" kind="lib" path="lib/common/RXTXcomm.jar">
 		<attributes>
 			<attribute name="javadoc_location" value="http://users.frii.com/jarvi/rxtx/doc/"/>
+			<attribute name="org.eclipse.jdt.launching.CLASSPATH_ATTR_LIBRARY_PATH_ENTRY" value="romraider/lib/windows"/>
```
Comment (dschultzca) [inline .classpath:26]: This is not required as the JVM command line parameter -Djava.library.path=lib/windows points to the proper location.

File: build.xml [hunk 2]
```diff
@@ -75,7 +75,7 @@
         <!-- java compiler properties -->
         <property name="javac.source" value="1.6" />
         <property name="javac.target" value="1.6" />
-        <property name="bootclasspath.dir" value="C:\Program Files (x86)\Java\jdk1.6.0_45\jre\lib" />
+       <property name="bootclasspath.dir" value="D:\Documents\EclipseProjects\RomRaider"/>
```
Comment (dschultzca) [inline build.xml:78]: This needs to resolve to the location of the Java 1.6 rt.jar, please remove.

File: src/main/java/com/romraider/logger/external/innovate/lm2/mts/plugin/Lm2MtsDataSource.java [hunk 3]
```diff
@@ -59,18 +62,21 @@
...
+	            for (int i = 0; i < ports.length; i++) {
+	                connector.usePort(i);
+	                dataItems = connector.getSensors();
+	            }
+	        }
+	        
+	        connector.dispose();
+    	}
+    	catch(com4j.ExecutionException e) {
+    		throw new PluginNotInstalledException("Innovate LogWorks MTS control does not appear to be installed on this computer");
```
Comment (dschultzca) [inline src/main/java/com/romraider/logger/external/innovate/lm2/mts/plugin/Lm2MtsDataSource.java:78]: Will this message be displayed to the user?  If so it needs to go into a locale properties file.

**Comment by dschultzca** [comment]
Add a baudrate attribute to transport if you need, but still need to be backwards compatible so don't remove it from the current element.

Why would there be a unknown transport protocol? It's required in the logger.xml.

If you could, please try to limit line length to about 80 characters where possible.  Scrolling left-right does not allow one to read the entire line of code.  Please refer to: https://github.com/RomRaider/RomRaider/wiki/RomRaider-Code-Style-Guide

File: build.xml (hunk 2 above)
Comment (GTO2013) [reply build.xml:78]: When I use the default build.xml I cannot compile it. I could compile it with other versions than 1.6 before, is 1.6 necessary for compilation now?

File: .classpath (hunk 1 above)
Comment (GTO2013) [reply .classpath:26]: Removed, this was unintentional actually

File: build.xml (hunk 2 above)
Comment (GTO2013) [reply build.xml:78]: Ah okay, so after some search this is needed so not only 1.6 rules are followed but also no new libraries are included

File: src/main/java/com/romraider/logger/external/innovate/lm2/mts/plugin/Lm2MtsDataSource.java (hunk 3 above)
Comment (GTO2013) [reply src/main/java/com/romraider/logger/external/innovate/lm2/mts/plugin/Lm2MtsDataSource.java:78]: It will be inside the log, but there will not be a window popup, neither will it be seen anywhere in the gui

**Comment by GTO2013** [comment]
I adjusted the line length to around 80 characters, also fixed the classpath and build.xml you mentioned. I tried to add some more attributes for the transport element, but I just ended up  reverting it, because the actual transport object doesnt seem to be easily accessible inside the connection. So it would be quite a mess... Defining the parameters in the id will have to do for now.

File: build.xml (hunk 2 above)
Comment (dschultzca) [reply build.xml]: Yes, without it and using Java 8 to cross compiling to 6, a few Java 8 references will creep in.  This is the default location for Java 6 JDK.  Adjust this locally for your own builds but I'd like to keep this default in the repo.

**Comment by dschultzca** [comment]
Thanks for the changes.
I have an Bluetooth OBDKey which presents a Serial port interface, should I expect your code to work with it?

**Comment by GTO2013** [comment]
If it uses an ELM327 inside it should be, yes. I dont currently have access to a car thats OBD-2 compiliant, but I will have soon. I only developed it on a simulator, so let me know if it works in an actual car (kinda important I guess?).
Also: I remember that I once built a bluetooth wideband logger for RR and I had to update the RXTX library locally to make it work. So if you run into a problem you can also try this.

**Comment by dschultzca** [comment]
Did you modify the logger.xml def to test this?

**Comment by GTO2013** [comment]
Yes and no, it will work with the stock OBD/CAN definition, because iso15765 will be converted to iso15765-4CAN11-500. But if you want to use another protocol you have to change the transport ID to one inside the list above.

File: src/main/java/com/romraider/io/connection/SerialConnectionProperties.java [hunk 4]
```diff
@@ -42,10 +42,11 @@ public int getBaudRate() {
         return baudRate;
     }
 
-    public void setBaudRate(int b) {
-
-    }
-
+	public void setBaudRate(int b) {
```
Comment (dschultzca) [inline src/main/java/com/romraider/io/connection/SerialConnectionProperties.java:45]: I don't see any references to this setter, can you restore this class?  This class is meant to be constructed with the appropriate values on creation.

**Comment by dschultzca** [comment]
> If it uses an ELM327 inside it should be, yes. 

So far my Bluetooth ELM adaptor doesn't play nice with RR Logger.  I believe it has to do with the initial connection to get the ECUID, then it closes the connection and re-opens to start logging.  For me, the COM port goes away between these steps.

**Comment by GTO2013** [comment]
Do you have a copy of the log maybe? What do you mean with goes away? The port disappears/disconnects?

**Comment by dschultzca** [comment]
> Do you have a copy of the log maybe? What do you mean with goes away? The port disappears/disconnects?

The port disappears and I can't even access it via PuTTY.  Not sure why it does this.  I captured one instance where it connected for a couple of seconds.  I doubt this is related to your code.
```
550401 INFO  [Thread-14] - Connected to: COM5
552990 INFO  [Thread-14] - Connection closed.
552990 DEBUG [Thread-14] - Error sending init
com.romraider.logger.ecu.exception.SerialCommunicationException: ELM327 was not found!
	at com.romraider.logger.ecu.comms.io.connection.ELMOBDLoggerConnection.ecuInit(ELMOBDLoggerConnection.java:70)
	at com.romraider.logger.ecu.comms.manager.QueryManagerImpl.doEcuInit(QueryManagerImpl.java:183)
	at com.romraider.logger.ecu.comms.manager.QueryManagerImpl.run(QueryManagerImpl.java:154)
	at java.lang.Thread.run(Thread.java:662)
```

**Comment by GTO2013** [comment]
Hmm, sounds pretty strange. Connected means receiving correct data from the car or just not crashing? I added some code to cycle through different baud rates on init, maybe that helps. Did you try the new RXTX version? I will have access to my car again soon, so I can do some testing myself

File: .classpath [hunk 5]
```diff
@@ -46,4 +46,4 @@
 	</classpathentry>
 	<classpathentry kind="con" path="org.eclipse.jdt.junit.JUNIT_CONTAINER/4"/>
 	<classpathentry kind="output" path="target"/>
```
Comment (dschultzca) [inline .classpath]: What's the reason for this change?

Comment (GTO2013) [reply .classpath]: If I remove the link to the native windows RXTX library eclipse doesnt find it. I always have to point it to the correct folder and then remove it later again before pushing. Not sure what is going on there

Comment (GTO2013) [reply .classpath]: So the native library paths are only added when doing an ANT build, so when you start it in eclipse directly its missing I guess?

Comment (dschultzca) [reply .classpath]: You need to set VM arguments in the Run Configuration for RomRaider.  Without this all native libs will not be found.  Just copy the line from the run.bat file.  Maybe that's why you have odd exceptions at startup.
image

Comment (GTO2013) [reply .classpath]: That solved it, thanks!

**Comment by dschultzca** [comment]
> Did you try the new RXTX version?

Which new version are you referring to?
I figured out a simple enumeration of ports via RXTX library causes the port (COM5) to appear then disappear.
Scanned every 1 second:
```
Stable Library
=========================================
Native lib Version = RXTX-2.1-7
Java lib Version   = RXTX-2.1-7
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
COM5 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------
COM1 - Serial
COM3 - Serial
COM4 - Serial
-----------------

```
